docs/develop/make_it_your_own.md

```diff
@@ -0,0 +1,443 @@
+    in invenio-rdm-records to know the keys for the other vocabularies.
+
+Restart your server and your vocabulary will now be used for resource types!
+
+
+## Extend the metadata model
+
+We've designed the default InvenioRDM metadata model to incorporate much of the
+useful fields the digital repository community has adopted over the years. From
+Subjects to Language to Location fields, there is a lot of depth to the out-of-the-box
+model. We encourage you to explore it fully before you consider adding your own
+fields. But adding your own fields is possible and no hacks are necessary.
+
+Metadata extensions are defined via two configurations: the additional fields'
+namespaces, a unique identifying url to prevent field clashes with other extensions,
+and the extensions (fields) themselves, a name, validation schema type and
+ElasticSearch storage type.
+
+To extend the metadata model, we edit the corresponding configuration variables
+in our familiar friend `invenio.cfg`. For our purposes, we first add the new
+namespace:
+
+
+```python
+# At the bottom
+RDM_RECORDS_METADATA_NAMESPACES = {
+    'dwc': {
+        # A URL ensures uniqueness (it doesn't *have* to resolve)
+        '@context': 'https://example.com/dwc/terms'
+    }
+}
+```
+
+New keys to `RDM_RECORDS_METADATA_NAMESPACES` (e.g. `'dwc'`) are shorthands for
+the unique `@context` values. They are used as namespace prefixes below.
+
+We then use the namespaces to prefix any field from that context in
+`RDM_RECORDS_METADATA_EXTENSIONS`, the dict of fields:
+
+```python
+# imports at the top...
+from marshmallow.fields import Bool
+
+from invenio_records_rest.schemas.fields import SanitizedUnicode
+
+# ...
+
+# At the bottom after RDM_RECORDS_METADATA_NAMESPACES above
+RDM_RECORDS_METADATA_EXTENSIONS = {
+    'dwc:family': {
+        'elasticsearch': 'keyword',
+        # You could make a field required if you wanted by using required=True
+        # e.g., SanitizedUnicode(required=True)
+        'marshmallow': SanitizedUnicode()
+    },
+    'dwc:behavior': {
+        'elasticsearch': 'text',
+        'marshmallow': SanitizedUnicode()
+    },
+    'dwc:right_or_wrong': {
+        'elasticsearch': 'boolean',
+        'marshmallow': Bool()
+    }
+}
+```
+
+Each key of `RDM_RECORDS_METADATA_EXTENSIONS` is of the form: `<prefix>:<field_key>`
+and each value a dict with the `'elasticsearch'` storage type and the `'marshmallow'`
+storage type. As of writing, the supported Elasticsearch storage types are:
+`'keyword'`, `'text'`, `'boolean'`, `'date'` and `'long'`. The supported
+Marshmallow schema types are:
+
+- `from marshmallow.fields`: `Bool`, `Integer`
+- `from invenio_records_rest.schemas.fields`: `DateString`, `SanitizedUnicode`
+
+and `marshmallow.fields.List` of the above.
+
+Restart the server. Creating a record now looks like this:
+
+```bash
+curl -k -XPOST -H "Content-Type: application/json" https://localhost:5000/api/records/ -d '{
```

Comment:
This `POST` opeartion would not work (at least in zsh firstly because the `False` instead of `false`.

In addition extensions are not returned:

``` bash
% curl -k -XGET https://localhost:5000/api/records/xj1xh-50n63 | jq .  
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1218  100  1218    0     0  17400      0 --:--:-- --:--:-- --:--:-- 17400
{
  "created": "2020-03-30T13:01:20.336555+00:00",
  "id": "xj1xh-50n63",
  "links": {
    "files": "https://localhost:5000/api/records/xj1xh-50n63/files",
    "self": "https://localhost:5000/api/records/xj1xh-50n63"
  },
  "metadata": {
    "_access": {
      "files_restricted": false,
      "metadata_restricted": false
    },
    "_created_by": 1,
    "_owners": [
      1
    ],
    "access_right": "open",
    "community": {
      "primary": "Maincom",
      "secondary": [
        "Subcom One",
        "Subcom Two"
      ]
    },
    "creators": [
      {
        "affiliations": [
          {
            "identifier": "entity-one",
            "name": "Entity One",
            "scheme": "entity-id-scheme"
          }
        ],
        "family_name": "Brutus",
        "given_name": "Marcus",
        "identifiers": {
          "Orcid": "9999-9999-9999-9990"
        },
        "name": "Marcus Junius Brutus",
        "type": "Personal"
      }
    ],
    "descriptions": [
      {
        "description": "A story about how permissions work.",
        "lang": "eng",
        "type": "Abstract"
      }
    ],
    "extensions": {},
    "identifiers": {
      "DOI": "10.9999/rdm.9999999",
      "arXiv": "9999.99999"
    },
    "licenses": [
      {
        "identifier": "BSD-3",
        "license": "Berkeley Software Distribution 3",
        "scheme": "BSD-3",
        "uri": "https://opensource.org/licenses/BSD-3-Clause"
      }
    ],
    "publication_date": "2020-03-30",
    "recid": "xj1xh-50n63",
    "resource_type": {
      "subtype": "image-photo",
      "type": "image"
    },
    "titles": [
      {
        "lang": "eng",
        "title": "A permission story",
        "type": "Other"
      }
    ]
  },
  "revision": 0,
  "updated": "2020-03-30T13:01:20.336559+00:00"
}
```

nor in search 

``` bash
curl -k -XGET "https://localhost:5000/api/records/?q=yarn" | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1587  100  1587    0     0  21160      0 --:--:-- --:--:-- --:--:-- 21160
{
  "aggregations": {
    "access_right": {
      "buckets": [
        {
          "doc_count": 1,
          "key": "open"
        }
      ],
      "doc_count_error_upper_bound": 0,
      "sum_other_doc_count": 0
    },
    "resource_type": {
      "buckets": [
        {
          "doc_count": 1,
          "key": "image"
        }
      ],
      "doc_count_error_upper_bound": 0,
      "sum_other_doc_count": 0
    }
  },
  "hits": {
    "hits": [
      {
        "created": "2020-03-30T13:01:20.336555+00:00",
        "id": "xj1xh-50n63",
        "links": {
          "files": "https://localhost:5000/api/records/xj1xh-50n63/files",
          "self": "https://localhost:5000/api/records/xj1xh-50n63"
        },
        "metadata": {
          "_access": {
            "files_restricted": false,
            "metadata_restricted": false
          },
          "_created_by": 1,
          "_owners": [
            1
          ],
          "access_right": "open",
          "community": {
            "primary": "Maincom",
            "secondary": [
              "Subcom One",
              "Subcom Two"
            ]
          },
          "creators": [
            {
              "affiliations": [
                {
                  "identifier": "entity-one",
                  "name": "Entity One",
                  "scheme": "entity-id-scheme"
                }
              ],
              "family_name": "Brutus",
              "given_name": "Marcus",
              "identifiers": {
                "Orcid": "9999-9999-9999-9990"
              },
              "name": "Marcus Junius Brutus",
              "type": "Personal"
            }
          ],
          "descriptions": [
            {
              "description": "A story about how permissions work.",
              "lang": "eng",
              "type": "Abstract"
            }
          ],
          "extensions": {},
          "identifiers": {
            "DOI": "10.9999/rdm.9999999",
            "arXiv": "9999.99999"
          },
          "licenses": [
            {
              "identifier": "BSD-3",
              "license": "Berkeley Software Distribution 3",
              "scheme": "BSD-3",
              "uri": "https://opensource.org/licenses/BSD-3-Clause"
            }
          ],
          "publication_date": "2020-03-30",
          "recid": "xj1xh-50n63",
          "resource_type": {
            "subtype": "image-photo",
            "type": "image"
          },
          "titles": [
            {
              "lang": "eng",
              "title": "A permission story",
              "type": "Other"
            }
          ]
        },
        "revision": 0,
        "updated": "2020-03-30T13:01:20.336559+00:00"
      }
    ],
    "total": 1
  },
  "links": {
    "self": "https://localhost:5000/api/records/?sort=bestmatch&q=yarn&size=10&page=1"
  }
}
```

However, as you can see the search triggered the record which means the content is there in ES. Seems to be marshmallow.

Reply:
Ah yes, the `False`s! I will investigate the extensions today.